hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/filtering/avro/AvroPathPredicate.java

```diff
@@ -35,31 +44,99 @@ public AvroPathPredicate(String path, Pattern pattern) {
+        ListIterator<String> iter = path.listIterator();
+        return select(record, iter);
+    }
+
+    private List<Object> select(GenericRecord record, ListIterator<String> iter) {
         Object current = record;
-        Iterator<String> iter = path.iterator();
         while (iter.hasNext()) {
             String selector = iter.next();
+            GenericRecord prevSelect = (GenericRecord) current;
             current = ((GenericRecord) current).get(selector);
+
+            if (current instanceof GenericRecord) {
+                continue;
+            }
+
+            Matcher arrayMatcher = ARRAY_PATTERN.matcher(selector);
+            if (arrayMatcher.matches()) {
+                String idx = arrayMatcher.group(GROUP_IDX);
+                selector = arrayMatcher.group(GROUP_SELECTOR);
+
+                if (idx.equals(WILDCARD_IDX)) {
+                    return selectMultipleArrayItems(iter, selector, prevSelect);
+                } else {
+                    current = selectSingleArrayItem(Integer.valueOf(idx), selector, prevSelect);
+                }
+
+            }
+
             if (!(current instanceof GenericRecord)) {
                 if (current == null) {
                     current = NULL_AS_STRING;
                 }
                 break;
             }
         }
-        return iter.hasNext() ? empty() : Optional.ofNullable(current);
+        return iter.hasNext() ? emptyList() : singletonList(current);
+    }
+
+    private List<Object> selectMultipleArrayItems(ListIterator<String> iter, String selector, GenericRecord prevSelect) {
+        Object current = prevSelect.get(selector);
+
+        if (!(current instanceof GenericArray)) {
+            return null;
+        }
+
+        GenericArray<Object> currentArray = ((GenericArray) current);
+        if (iter.hasNext()) {
+            return currentArray.stream()
+                .map(item -> {
+                    if (!(item instanceof GenericRecord)) {
+                        return emptyList();
+                    }
+                    return select((GenericRecord) item, path.listIterator(iter.nextIndex()));
+                })
+                .flatMap(List::stream)
+                .collect(Collectors.toList());
+        } else {
+            return currentArray.stream()
+                .map(item -> {
+                    if (item == null) {
+                        return NULL_AS_STRING;
+                    }
+                    return item;
+                })
+                .collect(Collectors.toList());
+        }
+    }
+
+    private Object selectSingleArrayItem(int idx, String selector, GenericRecord prevSelect) {
+        Object current = prevSelect.get(selector);
+
+        if (!(current instanceof GenericArray)) {
+            return null;
+        }
+
+        GenericArray currentArray = ((GenericArray) current);
+        if (currentArray.size() <= idx) {
+            return null;
```

Comment:
I would like to go back to your last comment from L131, but in this context.

Assume I have predicate like 
```json
{
  "path":  ".some.path.to.node", 
  "matcher": "123"
}
```
and my input message doesn't have record `path`. `AvroPathPredicate` will skip this message, right? Internally because `currentRecord.get(selector)` will return `null`. I would like to achieve same behaviour.

Maybe after my fixes to this PR it will be more clear.